xdsl/irdl/attributes.py

```diff
@@ -160,6 +164,7 @@ class ParamAttrDef:
 
     name: str
     parameters: list[tuple[str, AttrConstraint]]
+    converters: dict[str, Callable[[Any], Attribute]]
```

Comment:
I'm unsure if this is the correct place to store these. It's quite convenient though.

Reply:
Why not having a list here? Is there a reason we would need a `dict`?
Also, I think it would be nicer if the `parameters` would be something like a `ParamDef` class or something like this, rather than a `AttrConstraint`, so we can store the converter there directly.

Reply:
I guess a `dict` just seemed natural to me? Is there a good reason to have a list? The issue with storing a `ParamDef` in the `parameters` is that this function does processing to the constraint in the `ParamDef` (i.e. converting from type to constraint) so this would involve unbundling the original `ParamDef` doing the processing then rebundling it. What's in the PR was at least the cleanest way I could see to hold all the data but I'd be open to other suggestions (I tried having parameters be a "bigger" tuple with an optional converter and that seemed more messy, but I could try again and follow through with it if wanted)

Reply:
I mean, the reason I would do that is because `parameters` are also lists, so I don't think there is any reasons to have both a list and a dict there?
Also, what I meant is just having a `list[tuple[str, tuple[AttrConstraint, Callable]]` for parameters, and creating a class for that `tuple`. But that's fine by me if you want two fields (like you have ni the PR). But I would just make that `dict` a `list`.

Reply:
This feels worth spelling out, would you agree with these changes?

 1. make the value in the key/value pair list a ParamDef, so one level of indirection
 2. add an optional converter field to ParamDef

I actually considered proposing this and then thought that we are likely to have way fewer converters than other things, so having the ParamAttrDef have one dense structure and one sparse structure was probably the right way to go vs a dense structure of sparse structures. OTOH this is a bit more like dataclasses's Field, and it would be good for future extra features to have an obvious home...

Reply:
There's no such thing as a `ParamDef`. If you mean `_ParameterDef` then I disagree with the proposal, I think it's nice that the constraint field is actually a constraint by the time we get to this point.

For Mathieu, It would likely need to at least be a `list[tuple[str, tuple[AttrConstraint, Callable | None]]`. I really don't see what's wrong with the dictionary though tbh, we still have an ordering defined by the parameters field and this `converters` field feels like its just tagging on some extra info to me. I can change it to a list if really wanted, but just don't see the point?